[LTC] Sync gen_lazy_tensor.py from the staging branch

alanwaketan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #70385

Summary:
This commit sync gen_lazy_tensor.py from the lazy_tensor_staging branch
to the master.

Test Plan:
CI in the lazy_tensor_staging branch.

Differential Revision: [D33306232](https://our.internmc.facebook.com/intern/diff/D33306232)

Commit:
[LTC] Sync gen_lazy_tensor.py from the staging branch

Summary:
This commit sync gen_lazy_tensor.py from the lazy_tensor_staging branch
to the master.

Test Plan:
CI in the lazy_tensor_staging branch.

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33306232).

bdhirsh commented on tools/codegen/gen_lazy_tensor.py:
```diff
@@ -180,26 +181,6 @@ def gen_key(func: FunctionSchema) -> Tuple[str, str]:
         )),
     })
     # Generate headers for shape/dtype funcs for non-meta kernels
-    fm.write_with_template(f'{backend_key}ShapeInference.h', 'ShapeInference.h', lambda: {
```
Was this duplicated further down - that's why we're deleting it?

alanwaketan replied:
Yes, you are right.

wconstab commented on tools/codegen/gen_lazy_tensor.py:
```diff
@@ -180,26 +181,6 @@ def gen_key(func: FunctionSchema) -> Tuple[str, str]:
-            dest.GenLazyShapeInferenceDefinition(backend_indices[backend_key],
-                                                 tensor_class),
-            grouped_native_functions,
-            codegenInplaceVariant=True
```
but it's not quite a duplicate: the remaining copy does not have codegenInplaceVariant=True - are you sure which version is correct to keep?

alanwaketan replied:
The one on the branch does't have the `codegenInplaceVariant=True` which I assume is the latest one: https://github.com/pytorch/pytorch/blob/lazy_tensor_staging/tools/codegen/gen_lazy_tensor.py.

wconstab replied:
ok. i double checked, i think this is right.

alanwaketan commented:
Thanks, Will.

Commit:
Update on "[LTC] Sync gen_lazy_tensor.py from the staging branch"


Summary:
This commit sync gen_lazy_tensor.py from the lazy_tensor_staging branch
to the master.

Test Plan:
CI in the lazy_tensor_staging branch.

Differential Revision: [D33306232](https://our.internmc.facebook.com/intern/diff/D33306232)

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33306232).
